lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,448 @@
+                                      PlainStorage<T>>;
+// clang-format on
+
+namespace utils {
+// Utility functions for decoding arguments
+
+// Put a nullptr at the end of the array and poison it, this will make it so
+// out-of-bounds errors are caught quickly
+template<typename T>
+inline void poisonTail(std::vector<T *> &Vector) {
+  Vector.push_back(nullptr);
+  ASAN_POISON_MEMORY_REGION(Vector.back(), sizeof(T *));
+}
+
+inline std::vector<const char *> toStringList(ArgumentRef Argument) {
+  std::vector<const char *> Result;
+  std::vector<llvm::StringRef> ArgList = Argument.toStringList();
+  for (llvm::StringRef Arg : ArgList) {
+    Result.push_back(Arg.data());
+  }
+  poisonTail(Result);
+  return Result;
+}
+
+} // namespace utils
+
+static std::vector<string> TemporaryDirectories;
+static void cleanTemporaryDirectories(void *) {
+  for (auto &Directory : TemporaryDirectories) {
+    std::filesystem::remove_all(Directory);
+  }
+}
+
+class RunnerContext {
+private:
+  llvm::StringMap<uintptr_t> Pointers;
+
+public:
+  const revng::tracing::RunTraceOptions Options;
+
+public:
+  RunnerContext(const revng::tracing::RunTraceOptions Options) :
+    Options(Options) {}
+  ~RunnerContext() = default;
+  RunnerContext(const RunnerContext &Other) = delete;
+  RunnerContext(RunnerContext &&Other) = delete;
+  RunnerContext &operator=(const RunnerContext &Other) = delete;
+  RunnerContext &operator=(RunnerContext &&Other) = delete;
+
+  std::string getTemporaryDirectory() {
+    llvm::SmallString<128> TempOutput;
+    llvm::sys::fs::createUniqueDirectory("revng-run-trace", TempOutput);
+    std::string Output = TempOutput.str().str();
+    if (!Options.KeepTemporaries) {
+      TemporaryDirectories.push_back(Output);
+    } else {
+      dbg << "Creating temporary directory: " << Output << "\n";
+    }
+    return Output;
+  }
+
+  template<typename T>
+  void storePointer(const llvm::StringRef Name, T *Pointer) {
+    if (Name == NullPointer) {
+      return;
+    }
+    Pointers[Name] = reinterpret_cast<uintptr_t>(Pointer);
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_pointer_v<T>
+  T getPointer(const llvm::StringRef Name) {
+    // clang-format on
+    if (Name == NullPointer)
+      return nullptr;
+
+    revng_check(Pointers.count(Name) != 0,
+                "Tried to retrieve a pointer from context that was not "
+                "previously stored");
+    return reinterpret_cast<T>(Pointers[Name]);
+  }
+
+  void invalidatePointer(const llvm::StringRef Name) { Pointers.erase(Name); }
+
+  // clang-format off
+private:
+
+  template<ConstexprString Name, typename ArgT, size_t I>
+  requires (anyOf<ArgT, char *, const char *>())
+  std::vector<char> parseString(ArgumentRef Argument) {
+    // clang-format on
+    revng_check(Argument.isScalar(), "Argument is not scalar");
+    constexpr int LH = LengthHint<Name, I>;
+    ArgStorage<ArgT> Result;
```

Comment:
If you really really need it here, push `using ArgStorage` in here, but I'm sure you can at least prune one of the two `std::conditional_t`.